### PR TITLE
GDV: don't emit fallback call if classes are "exact"

### LINKED ISSUE #86235: Guarded devirtualization based on whole program view
In Native AOT we have a whole program view that often knows the full set of types that could be implementing an interface. RyuJIT already takes advantage of it for the case of interfaces implemented by exactly one class.

It would be nice if RyuJIT could also do guarded devirtualization for the case of multiple implementing classes. I'm looking at the ASP.NET Stage1 Goldilocks app profiles where I see we're spending about 3% of samples in `RhpInterfaceDispatch1`. This helper performs cached interface dispatch on monomorphic callsites and looks like this:

```
00007FF787DFFC80 4D 8B 5A 08          mov         r11,qword ptr [r10+8]  
00007FF787DFFC84 48 8B 01             mov         rax,qword ptr [rcx]  
00007FF787DFFC87 49 3B 43 20          cmp         rax,qword ptr [r11+20h]  
00007FF787DFFC8B 75 04                jne         Attempt1 (07FF787DFFC91h)  
00007FF787DFFC8D 41 FF 63 28          jmp         qword ptr [r11+28h]  
00007FF787DFFC91 E9 CA 07 00 00       jmp         RhpInterfaceDispatchSlow (07FF787E00460h)  
```

The fact that we see 3% of samples in these 6 assembly instructions means that:

1. Monomorphic interface dispatch is very common in ASP.NET
2. We're getting guaranteed CPU pipeline stalls very often: there's only one such helper shared for all interface calls and it's getting guaranteed mispredicted every time more than one interface call is in the hot path.

I had a look at the places where this happens and there's often only 2-3 classes that implement the interface in the whole program. It would likely help a ton if RyuJIT could generate inline if/else if/else if/else checks for these. It would be nice if the number of supported targets would be configurable through `--codegenopt` so that we can tune the right value of it for BLENDED/SPEED_OPT settings.

Cc @EgorBo

### PR DESCRIPTION
**Pull request by EgorBo**
Contributes to https://github.com/dotnet/runtime/issues/86769 and closes https://github.com/dotnet/runtime/issues/86235
```csharp
using System;
using System.Runtime.CompilerServices;
using System.Threading;

public interface IAnimal
{
    void MakeSound();
}

public class Cat : IAnimal
{
    public void MakeSound() => Console.WriteLine("Meow!");
}

public class Dog : IAnimal
{
    public void MakeSound() => Console.WriteLine("Woof!");
}

public class Cow : IAnimal
{
    public void MakeSound() => Console.WriteLine("Moo!");
}

// This class is not used and will be trimmed out
public class Lion : IAnimal
{
    public void MakeSound() => Console.WriteLine("Roar!");
}

public class Program
{
    public static void Main(string[] args)
    {
        Test(new Cat());
        Test(new Dog());
        Test(new Cow());
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Test(IAnimal a) => a.MakeSound();
}
```
NativeAOT codegen for `Test`:
```asm
; Method Program:Test(IAnimal)
       sub      rsp, 40
       lea      rax, [(reloc 0x4000000000423000)]
       cmp      qword ptr [rcx], rax                         ;; is it 'Cat'?
       jne      SHORT G_M31564_IG04
       lea      rcx, gword ptr [(reloc 0x4000000000423028)]  ;; '"Meow!"'
       call     System.Console:WriteLine(System.String)
       jmp      SHORT G_M31564_IG07
G_M31564_IG04:
       lea      rax, [(reloc 0x4000000000423010)]
       cmp      qword ptr [rcx], rax                         ;; is it 'Dog'?
       jne      SHORT G_M31564_IG06
       lea      rcx, gword ptr [(reloc 0x4000000000422f98)]  ;; '"Woof!"'
       call     System.Console:WriteLine(System.String)
       jmp      SHORT G_M31564_IG07
G_M31564_IG06:
       lea      rcx, gword ptr [(reloc 0x4000000000422f70)]  ;; '"Moo!"'
       call     System.Console:WriteLine(System.String)
G_M31564_IG07:
       nop      
       add      rsp, 40
       ret      
; Total bytes of code: 74
```
Virtual callback is not emitted since we took care of all possible classes implementing the given interface ("closed world" on NativeAOT). We might still emit it if we e.g. fail to inline one of the candidates (we currently don't do GDV just to devirtualize a call, we only do that when we're able to inline it) -- we might want to do that for cases where interface calls are expensive but that's a different task in https://github.com/dotnet/runtime/issues/86769

### CODE REVIEW

**Comment by EgorBo** [comment]
For reference, codegen for JIT:
```cs
static void Main()
{
    for (int i = 0; i < 200; i++)
    {
        Test(new Cat());
        Test(new Dog());
        Test(new Cow());
        Thread.Sleep(16);
    }
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Test(IAnimal a) => a.MakeSound();
```
```asm
; Assembly listing for method Program:Test(IAnimal)
       push     rsi
       sub      rsp, 32
       mov      rsi, qword ptr [rcx]
       mov      rax, 0x7FFF53C41C88      ; Cat
       cmp      rsi, rax
       jne      SHORT G_M31564_IG04
       mov      rcx, 0x26780208F80      ; 'Meow!'
       call     [System.Console:WriteLine(System.String)]
       jmp      SHORT G_M31564_IG07
G_M31564_IG04:
       mov      rax, 0x7FFF53C41E48      ; Dog
       cmp      rsi, rax
       jne      SHORT G_M31564_IG06
       mov      rcx, 0x26780209128      ; 'Woof!'
       call     [System.Console:WriteLine(System.String)]
       jmp      SHORT G_M31564_IG07
G_M31564_IG06:
       mov      rax, 0x7FFF53C42008      ; Cow
       cmp      rsi, rax
       jne      SHORT G_M31564_IG09
       mov      rcx, 0x26780209148      ; 'Moo!'
       call     [System.Console:WriteLine(System.String)]
G_M31564_IG07:
       nop      
       add      rsp, 32
       pop      rsi
       ret      
G_M31564_IG09:
       mov      r11, 0x7FFF53130280
       call     [r11]IAnimal:MakeSound():this
       jmp      SHORT G_M31564_IG07
```
(PGO-driven)

**Comment by EgorBo** [comment]
@AndyAyersMS PTAL, it passes NativeAOT outerloop tests + 3 type checks

File: src/coreclr/jit/likelyclass.cpp [hunk 1]
```diff
@@ -268,7 +273,7 @@ static unsigned getLikelyClassesOrMethods(LikelyClassMethodRecord*
 
                     // Distribute the rounding error and just apply it to the first entry.
                     // Assume that there is no error If we have unknown handles.
-                    if (numberOfClasses == h.m_totalCount)
+                    if (!containsUnknownHandles)
```
Comment (EgorBo) [inline src/coreclr/jit/likelyclass.cpp:276]: This is an unrelated change, it just that I used a wrong check in https://github.com/dotnet/runtime/pull/86965 so it didn't work properly

**Review by AndyAyersMS — APPROVED** [review]
